File: abac/tests/example_scripts/c/abac_prover.c
/**
   abac_prover.c

   To demonstrate how to use ABAC's api in C to make a query

   call:   abac_prover "keystorestring" "rolestring" "principalstring"

   pre-condition: run make attr_abac  generate IceCream_ID.pem and IceCream_private.pem with

   This program will make a prover call using 
           rolestring <- principalstring

**/

#include <err.h>
#include <stdio.h>
#include <assert.h>
#include <string.h>
#include <stdlib.h>

#include <abac.h>

int main(int argc, char **argv) {
    int i, success=0;
    abac_credential_t *cred=NULL;
    abac_credential_t **credentials=NULL;

    abac_context_t *ctx = abac_context_new();
    abac_context_load_directory(ctx, argv[1]);

    char *query=strdup(argv[2]);
    char *with=strdup(argv[3]);

    printf("query %s \n", query);
    printf("with %s\n", with);

int k=10; /* use to do repetitions */
while(k) {
    credentials = abac_context_query(ctx,
                            query, with,
                            &success);
    if (success)
        puts("prover success!!");
        else puts("prover failed!!");

    if (credentials != NULL && success) {
        puts("credentials needed :");
        for (i = 0; credentials[i] != NULL; ++i) {
           cred = credentials[i];
           printf("credential %s <- %s\n",
                    abac_role_string(abac_credential_head(cred)),
                    abac_role_string(abac_credential_tail(cred)));
        }
    }
    if(credentials)
        abac_context_credentials_free(credentials);
    k=k-1;
}
    free(query);
    free(with);

{ /* dump credentials from context */
    printf("\n\n");
    credentials = abac_context_credentials(ctx);
    printf("Dump context credentials, original ctx \n");
    if (credentials != NULL) {
        puts("context credentials :");
        for (i = 0; credentials[i] != NULL; ++i) {
           cred = credentials[i];
           printf("credential %s <- %s\n",
                    abac_role_string(abac_credential_head(cred)),
                    abac_role_string(abac_credential_tail(cred)));
        }
    }
    if(credentials)
        abac_context_credentials_free(credentials);
}

    abac_context_t *ctx2 = abac_context_dup(ctx);
{ /* dump credentials from context */
    printf("\n\n");
    credentials = abac_context_credentials(ctx);
    printf("Dump context credentials, original ctx2 \n");
    if (credentials != NULL) {
        puts("context credentials :");
        for (i = 0; credentials[i] != NULL; ++i) {
           cred = credentials[i];
           printf("credential %s <- %s\n",
                    abac_role_string(abac_credential_head(cred)),
                    abac_role_string(abac_credential_tail(cred)));
        }
    }
    if(credentials)
        abac_context_credentials_free(credentials);
}

    abac_context_free(ctx);
    abac_context_free(ctx2);

    return 0;
}

File: abac/tests/example_scripts/python/abac_prover.py
#!/usr/bin/env python

"""
   abac_prover.py
    
   To demonstrate how to use ABAC's api in python to make a query

   call:   abac_prover "keystorestring" "rolestring" "principalstring"

   pre-condition: run make attr_abac  generate IceCream_ID.pem and IceCream_private.pem with

   This program will make a prover call using
           rolestring <- principalstring

"""

import getopt
import sys
from ABAC import Context


def usage():
    print "Usage: abac_prover.py \\"
    print "        --keystore <keystore> \\"
    print "        --role <role> --principal <principal> "
    print "    loads the keystore and runs the query role <-?- principal"
    sys.exit(1)

keystore = ''
role = ''
principal = ''

try:
    opts, args = getopt.getopt(sys.argv[1:], '', ['keystore=', 'role=', 'principal='])
except getopt.GetoptError, err:
    print str(err)
    sys.exit(1)

for o, a in opts:
    if o == '--keystore':
        keystore = a
    elif o == '--role':
        role = a
    elif o == '--principal':
        principal = a
    else:
        assert False, "WAT"

if keystore == '' or role == '' or principal == '':
    usage()

# code begins here! sorry about that

print keystore
print role
print principal

ctx = Context()
ctx.load_directory(keystore)

(success, credentials) = ctx.query(role, principal)

if success:
    print "success"

for credential in credentials:
    print "credential %s <- %s" % (credential.head().string(), credential.tail().string())

